protocol_codegen/src/generate_messages/generate.rs

```diff
@@ -450,6 +404,50 @@ fn write_encode_field<W: Write>(
     Ok(())
 }
 
+fn write_encode_field_inner<W: Write>(
+    w: &mut CodeWriter<W>,
+    field: &PreparedField,
+    valid_versions: VersionSpec,
+    compute_size: bool,
+) -> Result<(), Error> {
+    let var_name = field.var_name();
+
+    let valid_versions = valid_versions.intersect(field.versions);
+    if !field.type_.has_compact_form() {
+        write_encode_or_compute(w, &field.type_.name(false), &var_name, compute_size)
+    } else {
+        write_version_cond(
+            w,
+            valid_versions,
+            field.flexible_versions,
+            |w| write_encode_or_compute(w, &field.type_.name(true), &var_name, compute_size),
+            |w| write_encode_or_compute(w, &field.type_.name(false), &var_name, compute_size),
+            false,
+            false,
+        )
+    }
+}
+
+fn write_default_check<W: Write>(
+    w: &mut CodeWriter<W>,
+    field: &PreparedField,
+) -> Result<(), Error> {
+    let var_name = field.var_name();
+
+    write!(
+        w,
+        "if {} ",
+        field
+            .default
+            .gen_is_default(&var_name, field.optional)
+            .not()
+    )?;
+    w.block(|w| {
+        write!(w, "return Err(EncodeError)")?;
+        Ok(())
+    })
```

Comment:
I think some of these could still be improved by lifting vars up, to ensure that the `write!` call is formatted as a single line, which can help comprehension of what the generated code looks like:

```suggestion
    let var_name = field.var_name();
    let field = field
        .default
        .gen_is_default(&var_name, field.optional)
        .not();

    write!(w, "if {} ", field)?;
    w.block(|w| {
        write!(w, "return Err(EncodeError)")?;
        Ok(())
    })
```